src/effects/effectparameterslot.cpp

```diff
@@ -56,62 +56,70 @@ EffectParameterSlot::~EffectParameterSlot() {
     delete m_pSoftTakeover;
 }
 
-void EffectParameterSlot::loadEffect(EffectSlot* pEffectSlot) {
-    // qDebug() << debugString() << "loadEffect" << (pEffectSlot ? pEffectSlot->getManifest().name() : "(null)");
+void EffectParameterSlot::updateEngineState() {
```

Comment:
No, we cannot move this to EffectParameterSlotBase because updateEngineState function uses `m_pControlValue`. We cannot declare `m_pControlValue` in EffectParameterSlotBase because it's datatype depends on whether it is knob parameter or button parameter. We can make a `EffectParameterSlotBase::updateEngineState(double value)` method and declare wrapper functions in EffectParameterSlot and EffectButtonParameterSlot.

Reply:
Are you sure? `m_pControlValue` is a subclass of `ControlObject` in both `EffectParameterSlotBase` subclasses, so I think you can declare it with the `ControlObject` type in `EffectParameterSlotBase` and initialize it as nullptr in the `EffectParameterSlotBase` member initialization list. Then the `EffectParameterSlotBase` subclass constructors could initialize `m_pControlValue` with the appropriate `ControlObject` subclass and make the Qt signal-slot connection.